rooms.js

```diff
@@ -163,6 +165,109 @@ var Room = (function () {
 		}
 		return true;
 	};
+	//mute handling
+	Room.prototype.runMuteTimer = function () {
+		if (this.muteTimer || this.muteQueue.length === 0) return;
+
+		var timeUntilExpire = this.muteQueue[0].time - Date.now();
+		if (timeUntilExpire <= 0) {
+			this.unmute(this.muteQueue[0].userid, true);
+			//runMuteTimer() is called again in unmute() so this function instance should be closed
+			return;
+		}
+		var self = this;
+		this.muteTimer = setTimeout(function () {
+			self.muteTimer = null;
+			self.runMuteTimer();
+		}, timeUntilExpire);
+	};
+	Room.prototype.isMuted = function (user) {
+		if (!user) return;
+		if (this.muteQueue) {
+			for (var i = 0; i < this.muteQueue.length; i++) {
+				if (user.userid === this.muteQueue[i].userid ||
+					user.guestNum === this.muteQueue[i].guestNum ||
+					(user.autoconfirmed && user.autoconfirmed === this.muteQueue[i].autoconfirmed)) {
+					return this.muteQueue[i].userid;
+				}
+			}
+		}
+	};
+	Room.prototype.getMuteTime = function (user) {
+		var userid = this.isMuted(user);
+		if (!userid) return;
+		for (var i = 0; i < this.muteQueue.length; i++) {
+			if (userid === this.muteQueue[i].userid) {
+				return this.muteQueue[i].time - Date.now();
+			}
+		}
+	};
+	Room.prototype.mute = function (user, setTime) {
+		var userid = user.userid;
+
+		if (!setTime) setTime = 7 * 60000; // default time: 7 minutes
+		if (setTime > 90 * 60000) setTime = 90 * 60000; // limit 90 minutes
+
+		// If the user is already muted, the existing queue position for them should be removed
+		if (this.isMuted(user)) this.unmute(userid);
+
+		// Place the user in a queue for the unmute timer
+		for (var i = 0; i <= this.muteQueue.length; i++) {
+			var time = Date.now() + setTime;
+			if (i === this.muteQueue.length || time < this.muteQueue[i].time) {
+				var entry = {
+					userid: userid,
+					time: time,
+					guestNum: user.guestNum,
+					autoconfirmed: user.autoconfirmed
+				};
+				this.muteQueue.splice(i, 0, entry);
+				// The timer needs to be switched to the new entry if it is to be unmuted
+				// before the entry the timer is currently running for
+				if (i === 0 && this.muteTimer) {
+					clearTimeout(this.muteTimer);
+					this.muteTimer = null;
+				}
+				break;
+			}
+		}
+		this.runMuteTimer();
+
+		user.updateIdentity(this.id);
+		return userid;
+	};
+	Room.prototype.unmute = function (userid, sendPopup) {
+		var user = Users(userid);
+		if (!user) {
+			//if the user cannot be found, construct a dummy object for them
```

Comment:
SOP is just to use the `userid` variable - check the unlock code for an example. You don't need a separate set of logic because that's the only set of logic you need.